File: yahtzee.py
import random
import statistics


def simulate_yahtzee(n_dice=5, tries=3):
    dice = [random.randint(1, 6) for _ in range(n_dice)]
    for _ in range(tries):
        if len(set(dice)) == 1:
            return True
        most_common = statistics.mode(dice)
        dice = [
            x if x == most_common else random.randint(1, 6)
            for x in dice
        ]
    return False


if __name__ == "__main__":
    n = 1000000
    n_wins = 0
    for i in range(n):
        if simulate_yahtzee():
            n_wins += 1
        if (i+1) % (n//10) == 0:
            print(f"(i={i}) Percentage: {n_wins / (i+1)}")

File: yahtzee.js
function mode(l) {
  const m = new Map();
  for (const x of l) {
    if (m.has(x)) {
      m.set(x, m.get(x) + 1);
    } else {
      m.set(x, 1);
    }
  }
  let mostCommon = null;
  let mostCommonCount = 0;
  for (const [k, v] of m.entries()) {
    if (v > mostCommonCount) {
      mostCommonCount = v;
      mostCommon = k;
    }
  }
  return mostCommon;
}

function simulateYahtzee(nDice = 5, tries = 3) {
  let dice = Array(nDice).fill(null).map(_ => Math.floor(6 * Math.random()) + 1);
  for (let i = 0; i < tries; i++) {
    const s = new Set(dice);
    if (s.size == 1) {
      return true;
    }
    mostCommon = mode(dice);
    dice = dice.map(x => x == mostCommon ? x : Math.floor(6 * Math.random()) + 1);
  }
  return false;
}

const n = 100000;
let nWins = 0;
for (let i = 0; i < n; i++) {
  nWins += !!simulateYahtzee();
  if ((i + 1) % (n / 10) == 0) {
    console.log(`(i=${i}) Percentage: ${nWins / (i + 1)}`);
  }
}
